File: src/main/java/pcb/units/concrete/time/Hours.java
package pcb.units.concrete.time;

import pcb.units.amount.BigDecimalAmount;
import pcb.units.base.AbstractUnitAmount;
import pcb.units.base.UnitAmount;
import pcb.units.dimensions.time.Time;
import pcb.units.dimensions.time.TimeUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class Hours
		extends AbstractUnitAmount<BigDecimalAmount, TimeUnit>
		implements Time {

	// region constructors

	public Hours(Number value) {
		this(value.toString());
	}

	public Hours(String value) {
		this(new BigDecimal(value));
	}

	public Hours(BigDecimal value) {
		this(new BigDecimalAmount(value));
	}

	public Hours(BigDecimalAmount amount) {
		super(amount, HourUnit.instance);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Hours plus(UnitAmount<BigDecimalAmount, TimeUnit> other, MathContext mathContext) {
		Hours meters = new Hours(other.getAmountIn(getUnit()));
		return new Hours(getAmount().plus(meters.getAmount(), mathContext));
	}

	@Override
	public Hours minus(UnitAmount<BigDecimalAmount, TimeUnit> other, MathContext mathContext) {
		Hours meters = new Hours(other.getAmountIn(getUnit()));
		return new Hours(getAmount().minus(meters.getAmount(), mathContext));
	}

	@Override
	public Hours multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Hours(getAmount().multipliedBy(other, mathContext));
	}

	@Override
	public Hours dividedBy(BigDecimal other, MathContext mathContext) {
		return new Hours(getAmount().dividedBy(other, mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/base/Hours.kt
package pcb.uwu.amount.base

import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.HourUnit
import pcb.uwu.unit.base.TimeUnit

class Hours : Time
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, HourUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, HourUnit)

    // region UnitAmount

    override fun plus(time: UnitAmount<TimeUnit>) =
        Hours(this.amount + (time to HourUnit).amount)

    override fun minus(time: UnitAmount<TimeUnit>) =
        Hours(this.amount - (time to HourUnit).amount)

    override fun times(scalar: Number) =
        Hours(this.amount * scalar)

    override fun div(scalar: Number) =
        Hours(this.amount / scalar)

    // endregion
}
